Apps/TimelineApp/feature/user/src/main/java/uk/ac/cam/cares/jps/user/SensorSettingFragment.java

```diff
@@ -127,72 +138,88 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
+        binding.toggleAllBtn.setEnabled(false);
+    }
+
     /**
      * Resets the state of the "Toggle All" button to its default state.
      */
     private void resetToggleAllButton() {
         binding.toggleAllBtn.setText(R.string.toggle_all);
     }
 
-    /**
-     * Updates the state of the "Start Recording" button based on the current sensor toggle states.
-     */
-    private void updateStartRecordingButtonState() {
-        // enable the start recording button only if at least one sensor is toggled on
-        boolean hasToggledOnSensor = false;
-        for (SensorItem item : adapter.getSensorItems()) {
-            if (item.isToggled()) {
-                hasToggledOnSensor = true;
-                break;
-            }
-        }
-        binding.startRecordTv.setEnabled(hasToggledOnSensor);
-    }
 
     /**
      * Handles the logic for starting or stopping sensor recording when the button is clicked.
      *
      * @param view The view that was clicked.
      */
     private void onRecordButtonClicked(View view) {
-        // check permission
-        for (Permission permission : permissionsMap.values()) {
-            checkFineLocationPermission(permission);
+        boolean locationToggled = false;
+        boolean audioToggled = false;
+
+        List<SensorType> selectedSensorTypes = new ArrayList<>();
+
+        for (SensorItem item : adapter.getSensorItems()) {
+            if (item.isToggled()) {
+                SensorType sensorType = item.getSensorType();
+                selectedSensorTypes.add(sensorType);
+
+                if (sensorType == SensorType.LOCATION) {
+                    locationToggled = true;
+                } else if (sensorType == SensorType.SOUND) {
+                    audioToggled = true;
+                }
+            }
         }
 
-        String permissionNotGranted = permissionsMap.values().stream()
-                .filter(permission -> criticalPermissionType.contains(permission.type) && !permission.isGranted)
-                .map(permission -> permission.type.toString())
-                .collect(Collectors.joining(" "));
-        if (!permissionNotGranted.isEmpty()) {
-            Toast.makeText(requireContext(), permissionNotGranted + " not granted. Not able to start recording.", Toast.LENGTH_LONG).show();
+        if (selectedSensorTypes.isEmpty()) {
+            Toast.makeText(requireContext(), "Please enable at least one sensor or click Toggle All", Toast.LENGTH_SHORT).show();
             return;
         }
 
-        // Gather selected sensor types
-        List<SensorItem> selectedSensorTypes = new ArrayList<>();
-        for (SensorItem item : adapter.getSensorItems()) {
-            if (item.isToggled()) {
-                selectedSensorTypes.add(item);
+        Runnable startRecordingRunnable = () -> startRecording(selectedSensorTypes);
+
+        if (locationToggled) {
+            Permission permission = permissionsMap.get(Permission.PermissionType.LOCATION_FINE);
```

Comment:
Since now we are checking the permission only when the button has been toggled, there is no need to use `permissionMap` any more. Can just create the permission instance when needed